### PR TITLE
fix(core): saturate UPGDLearner step_count at int32 max

### LINKED ISSUE #1841: upgd: UPGDLearner.update step_count wraps negative at int32 max
## Summary

`UPGDLearner.update` advances its lifetime `step_count` with a bare
`state.step_count + 1` (`alberta_framework/core/upgd.py:3330`). The counter
is declared `dtype=jnp.int32` (`upgd.py:1781`) and there is no saturating
increment anywhere in the file. At `step_count == 2**31 - 1` the next update
wraps the clock to `INT32_MIN` instead of saturating at `INT32_MAX`.

`step_count` drives the warm-up, meta-warm, and perturbation schedule decisions
inside `update` (`upgd.py:2393`, `2418`, `2873-2876`, `3207`, `3242`). A
wrapped negative clock flips `warm`/`after_warmup`/`meta_warm` to `False`
permanently and changes which updates the perturbation schedule classifies,
so lifetime-exhausted learners no longer reproduce their pre-wrap behavior.

This is the same counter-wrap class already fixed elsewhere in the tree
(`learners.py` #1372, `synthetic.py` #1366, `feature_discovery.py` #1368,
out-of-class #1441, `horde.py` #1787, `reward_model.py` #1017); UPGD was not
covered.

## Reproduction

```python
import jax.numpy as jnp
import jax.random as jr
from alberta_framework.core.upgd import UPGDLearner

learner = UPGDLearner(
    n_heads=1, hidden_sizes=(4,), sparsity=0.0, perturbation_sigma=0.0
)
state = learner.init(feature_dim=4, key=jr.key(0)).replace(
    step_count=jnp.asarray(2**31 - 1, dtype=jnp.int32)
)
result = learner.update(
    state, jnp.zeros(4, dtype=jnp.float32), jnp.zeros(1, dtype=jnp.float32)
)
print(int(result.state.step_count))  # -2147483648 (wrap), expected 2147483647
```

`before: 2147483647 after: -2147483648` — the clock wraps negative instead of
saturating.

## Impact

- `step_count` is a lifetime identity field: `UPGDState` stores it as the
  scalar `step_count` telemetry and downstream warm/perturbation branches read
  it. Wrapping to a negative value silently changes every `>= warmup_steps`
  comparison and the `% perturbation_interval` schedule for the rest of the
  life.
- The fix matches the tree-wide convention: saturate the counter at
  `INT32_MAX` using the same guard used by the other lifetime-clock fixes.

## Fix

Replace the bare increment at `upgd.py:3330` with a saturating int32 increment
(floor at 0, cap at `INT32_MAX - 1`, then `+ 1`), consistent with
`normalizers._saturating_int32_counter_increment` and the other core modules.
Add a regression test that plants `step_count == INT32_MAX`, runs one update,
and asserts the clock saturates (and stays non-negative) rather than wrapping.

Fix incoming in a linked PR with the reproduction test.

### PR DESCRIPTION
**Pull request by deepanshu-yd**
Closes #1841

## Problem

`UPGDLearner.update` advanced the lifetime `step_count` with a bare `state.step_count + 1` (`alberta_framework/core/upgd.py:3330`). The counter is declared `dtype=jnp.int32` and has no saturating increment. At `step_count == 2**31 - 1` the next update wraps the clock to `INT32_MIN`.

The wrapped negative clock flips the warm-up, meta-warm, and perturbation-schedule branches inside `update` (`upgd.py:2393`, `2418`, `2873-2876`, `3207`, `3242`) to `False` permanently, so an exhausted learner no longer reproduces its pre-wrap update behavior.

## Fix

Replace the bare increment with the tree-wide saturating int32 counter increment (`normalizers._saturating_int32_counter_increment`), matching the existing lifetime-clock fixes in learners, synthetic streams, feature discovery, horde, and reward_model.

## Tests

Adds `tests/test_upgd_lifetime_clock.py`: plants the exhausted clock and asserts the counter saturates at `INT32_MAX` and never wraps negative.

- `pytest tests/test_upgd.py tests/test_upgd_lifetime_clock.py`: 94 passed
- `ruff check`: clean
- `mypy`: clean

### CODE REVIEW

**Comment by lalalune** [comment]
Incorporated and validated in #1874, which preserved this contribution in the consolidated history and has now merged to main. Closing this source PR as resolved.
